### ISSUE TITLE
[DataGrid] Virtualization not respecting scrollTop

### ISSUE THREAD

**Issue by oliviertassinari** [mui-org/material-ui-x#1911]
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

End-users can feel any difference between virtualization enable and disabled

## Expected Behavior 🤔

End-users shouldn't feel any difference between virtualization enable and disabled

## Steps to Reproduce 🕹

Steps:

1. Open https://codesandbox.io/s/material-uix-gridscroll-to-indexes-demo-simple-forked-on0dc
2. See how the scroll-top is **discontinue** disorienting end-users.
![scrollTop](https://user-images.githubusercontent.com/3165635/122136528-f9db6300-ce42-11eb-8f4c-37e43005ff17.gif)

You can also see there is a blank space at the bottom, but it's #414.

## Context 🔦

Raised by @m4theushw in #1871.

>  I found that the amount of scroll needed to make a row visible is less than what I would normally expect (rowIndex * rowHeight)

It's KO. It would only make sense for this use case: https://ag-grid.com/react-grid/massive-row-count/.

I suspect the best course of action is to rewrite the virtualization logic. Taking the following constraints into account in the design.

- Built-in support for no virtualization #1781
- Variable row height #417
- Dynamic row height #438
- Performance #1903
- Research into overscan #270

**Comment by oliviertassinari** [comment]
Looking at the reproduction, we are completely off. We have 34 rows of 50px height with a viewport of 290px height. So the max scroll top should be = 34 * 50 - 290 = 1410px. Instead, it's 1250px.

**Comment by m4theushw** [comment]
> Looking at the reproduction, we are completely off. We have 34 rows of 50px height with a viewport of 290px height. So the max scroll top should be = 34 * 50 - 290 = 1410px. Instead, it's 1250px 🙃.

Yeah, when I was playing with the virtualization to understand it I noticed that we use a different calculation for the height of the container. If we do the right thing the user will have to scroll more, but there will be no discontinuation when reaching the top.

There's also a blank space that might appear when calling `scrollToIndexes`. Making the scrollTop to honor the theoretical position of the row will fix it too. BTW, we could improve this API to allow the user to determine if he wants: 1. scroll as little as possible to make the row visible or 2. scroll as much necessary so the row is at the top of the viewport. AG Grid in `ensureIndexVisible` even allows to scroll to place the row at the middle or bottom of the viewport.

<img width="500" src="https://user-images.githubusercontent.com/42154031/122138385-04830180-ce1d-11eb-9687-cd3d269956fe.png" alt="image" style="max-width:100%;">

**Comment by oliviertassinari** [comment]
> we could improve this API to allow the user to determine if he wants: 1. scroll as little as possible to make the row visible or 2. scroll as much necessary so the row is at the top of the viewport. AG Grid in ensureIndexVisible even allows to scroll to place the row at the middle or bottom of the viewport.

True, makes me think about the different options of https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView
